app/src/main/java/com/youssef/fixit/Auth/LoginFragment/LoginPresenter.java

```diff
@@ -46,16 +47,17 @@ public void onResponse(Call<Register> call, Response<Register> response) {
                 view.hideLoading();
                 if (response.isSuccessful() && response.code() == 200) {
                     if (response.body().getData() != null) {
+                        Data data = response.body().getData();
                         List<String> roles = response.body().getData().getRoles();
                         SplashScreen.MyToken = response.body().getData().getToken();
                         SplashScreen.My_ID = response.body().getData().getId();
                         SplashScreen.MyRole = roles.get(0);
                         Log.d(TAG, response.body().getData().getId() + "");
                         Log.d(TAG, SplashScreen.MyToken);
-                        String token = response.body().getData().getToken();
-                        String role = response.body().getData().getRoles().get(0);
-                        int my_id = response.body().getData().getId();
-                        view.onLoginSuccessful(token, role, my_id);
+                        view.saveString("token", data.getToken());
```

Comment:
One unrelated note here, always try to cleanup resources. This will be achieved when you know when the presenter will attached and when it will be destroyed. And since the presenter is bound to an activity with a lifecycle you  can use that. 

So it will go like that:
Actiivty#onDestroy will trigger presenter#onDestroy here you can release the view and nullfiy it. 
it will be something like that in the presenter
```
void onDestroy() {
view = null;
}
```

Why this is needed because right now the presenter will have for ever a reference for a dead view and it will accumulate eventually to memory leaks. beside that it might cause crashes as well. for an example say the user tried to login and the login request initiated but before receiving any result the user closed the activity/fragment therefore all the views' references have been destroyed and no longer valid. therefore, the application will crash.

So it's always save to check if the view still exists and this by binding the presenter to the activity/fragment (view) lifecycle.